README.md
# class-sanitizer

![Build Status](https://github.com/typestack/class-sanitizer/workflows/CI/badge.svg)
[![codecov](https://codecov.io/gh/typestack/class-sanitizer/branch/master/graph/badge.svg)](https://codecov.io/gh/typestack/class-sanitizer)
[![npm version](https://badge.fury.io/js/class-sanitizer.svg)](https://badge.fury.io/js/class-sanitizer)

Decorator based class property sanitation in Typescript powered by [validator.js][validator.js].

> **DEPRECATION NOTICE:**  
> This library is considered to be deprecated and won't be updated anymore. Please use the [class-transformer][ct] and/or [class-validator][cv] libraries instead.

## Installation

```bash
npm install class-sanitizer --save
```

## Usage

To start using the library simply create some classes and add some sanitization decorators to the properties. When calling
`sanitize(instance)` the library will automatically apply the rules defined in the decorators to the properties and update
the value of every marked property respectively.

> **NOTE:**  
> Every sanitization decorator is property decorator meaning it cannot be placed on parameters or class definitions.

```typescript
import { sanitize, Trim } from 'class-sanitizer';

class TestClass {
  @Trim()
  label: string;

  constructor(label: string) {
    this.label = label;
  }
}

const instance = new TestClass(' text-with-spaces-on-both-end ');

sanitize(instance);
// -> the label property is trimmed now
// -> { label: 'text-with-spaces-on-both-end' }
```

### Validating arrays

Every decorator expects a `SanitationOptions` object. When the `each` property is set to `true`
the array will be iterated and the decorator will be applied to every element of the array.

```ts
import { sanitize, Trim } from 'class-sanitizer';

class TestClass {
  @Trim(undefined, { each: true })
  labels: string[];

  constructor(labels: string[]) {
    this.labels = labels;
  }
}

const instance = new TestClass([' labelA ', ' labelB', 'labelC ']);

sanitize(instance);
// -> Every value is trimmed in instance.labels now.
// -> { labels: ['labelA', 'labelB', 'labelC']}
```

### Inheritance

Class inheritance is supported, every decorator defined on the base-class will
be applied to the property with same name on the descendant class if the property exists.

> **Note**:  
> **Only one level of inheritance is supported!** So if you have `ClassA` inherit `ClassB` which inherits `ClassC` the
> decorators from `ClassC` won't be applied to `ClassA` when sanitizing.

```ts
import { sanitize, Trim } from 'class-sanitizer';

class BaseClass {
  @Trim()
  baseText: string;
}

class DescendantClass extends BaseClass {
  @Trim()
  descendantText: string;
}

const instance = new DescendantClass();
instance.baseText = ' text ';
instance.descendantText = ' text ';

sanitize(instance);
// -> Both value is trimmed now.
// -> { baseText: 'text', descendantText: 'text' }
```

### Sanitizing nested values with `@SanitizeNested()` decorator

The `@SanitizeNested` property can be used to instruct the library to lookup the sanitization rules
for the class instance found on the marked property and sanitize it.

```ts
import { sanitize, Trim, SanitizeNested } from 'class-sanitizer';

class InnerTestClass {
  @Trim()
  text: string;

  constructor(text: string) {
    this.text = text;
  }
}

class TestClass {
  @SanitizeNested({ each: true })
  children: InnerTestClass[];

  @SanitizeNested({ each: false })
  child: InnerTestClass;
}

const instance = new TestClass();
const innerA = new InnerTestClass(' innerA ');
const innerB = new InnerTestClass(' innerB ');
const innerC = new InnerTestClass(' innerC ');
instance.children = [innerA, innerB];
instance.child = innerC;

sanitize(instance);
// -> Both values in the array on `children` property and value on `child` property is sanitized.
// -> { children: [ { text: 'innerA' }, { text: 'innerB' }], child: { 'innerC' }}
```

### Custom sanitation classes

The `@SanitizerConstraint(` decorator can be used to define custom sanitization logic. Creating a custom sanitization class requires the following steps:

1. Create a class which implements the `CustomSanitizer` interface and decorate the class with the `@SanitizerConstraint()` decorator.

   ```typescript
   import { CustomSanitizer, SanitizerConstraint } from 'class-sanitizer';
   import { Container } from 'typedi';

   @SanitizerConstraint()
   export class LetterReplacer implements CustomSanitizer {
     /** If you use TypeDI, you can inject services to properties with `Container.get` function. */
     someInjectedService = Container.get(SomeClass);

     /**
      * This function will be called during sanitization.
      *  1, It must be a sync function
      *  2, It must return the transformed value.
      */

     sanitize(text: string): string {
       return text.replace(/o/g, 'w');
     }
   }
   ```

1. Then you can use your new sanitation constraint in your class:

   ```typescript
   import { Sanitize } from 'class-sanitizer';
   import { LetterReplacer } from './LetterReplacer';

   export class Post {
     @Sanitize(LetterReplacer)
     title: string;
   }
   ```

1. Now you can use sanitizer as usual:

   ```typescript
   import { sanitize } from 'class-sanitizer';

   sanitize(post);
   ```

### Manual sanitation

There are several method exist in the Sanitizer that allows to perform non-decorator based sanitation:

```typescript
import Sanitizer from 'class-sanitizer';

Sanitizer.trim(` Let's trim this! `);
```

## Sanitization decorators

The following property decorators are available.

| Decorator                              | Description                                                                                                                              |
| -------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------- |
| `@Blacklist(chars: string)`            | Removes all characters that appear in the blacklist.                                                                                     |
| `@Whitelist(chars: string)`            | Removes all characters that don't appear in the whitelist.                                                                               |
| `@Trim(chars?: string)`                | Trims characters (whitespace by default) from both sides of the input. You can specify chars that should be trimmed.                     |
| `@Ltrim(chars?: string)`               | Trims characters from the left-side of the input.                                                                                        |
| `@Rtrim(chars?: string)`               | Trims characters from the right-side of the input.                                                                                       |
| `@Escape()`                            | Replaces <, >, &, ', " and / with HTML entities.                                                                                         |
| `@NormalizeEmail(lowercase?: boolean)` | Normalizes an email address.                                                                                                             |
| `@StripLow(keepNewLines?: boolean)`    | Removes characters with a numerical value < 32 and 127, mostly control characters.                                                       |
| `@ToBoolean(isStrict?: boolean)`       | Converts the input to a boolean. Everything except for '0', 'false' and '' returns true. In strict mode only '1' and 'true' return true. |
| `@ToDate()`                            | Converts the input to a date, or null if the input is not a date.                                                                        |
| `@ToFloat()`                           | Converts the input to a float, or NaN if the input is not an integer.                                                                    |
| `@ToInt(radix?: number)`               | Converts the input to an integer, or NaN if the input is not an integer.                                                                 |
| `@ToString()`                          | Converts the input to a string.                                                                                                          |

[validator.js]: https://github.com/chriso/validator.js
[typedi]: https://github.com/pleerock/typedi
[ct]: https://github.com/typestack/class-transformer
[cv]: https://github.com/typestack/class-validator


CHANGELOG.md
# Changelog

_This changelog follows the [keep a changelog][keep-a-changelog]_ format to maintain a human readable changelog.

## [1.0.1][v1.0.1] - 2020-07-31

> This is the **final release of the library**, if you lack some functionality please use [class-transformer][ct] and [class-validator][cv] instead.

#### Fixed

- sanitization won't skip properties with empty string values anymore (`''`)

## [1.0.0][v1.0.0] - 2020-07-31 [BREAKING CHANGE]

#### Fixed

- passing `{ each: true }` into decorators don't crashes the library anymore
- fixed security warning due to out-of-date dev dependencies

#### Added

- added support for `@SanitizeNested()` operator
- added unit-test for the whole project
- inheritance now works as expected, base properties will be sanitized on the descendant class

#### Changed

- renamed `SanitizerInterface` to `CustomSanitizer` [BREAKING]
- removed deprecated entry-point for the project [BREAKING]
- updated validator.js to latest
- improved metadata lookup speed via storing them in `Map` instead of an array
- updated project tooling (Typescript, Eslint, Prettier, Jest, GH Actions)
- greatly improved the readability of the code
- restructured project to match the structure of our other projects

## [0.0.5][v0.0.5] - 2017-03-06

#### Fixed

- published version contains all the files instead of just `package.json` and `README.md` files

## [0.0.4][v0.0.4] - 2017-03-03

#### Fixed

- when the correct type is passed to a decorator it will be returned immediately instead of passing it to validator.js and breaking it

## [0.0.3][v0.0.3] - 2017-01-15

#### Fixed

- added `index.ts` entrypoint to allow importing from package-root
- fixed inheritance support, decorators are now respected from the base-class

## 0.0.1 - 2016-04-18

Initial version.

[v1.0.0]: https://github.com/typestack/class-sanitizer/compare/v1.0.0...v1.0.1
[v1.0.0]: https://github.com/typestack/class-sanitizer/compare/v0.0.5...v1.0.0
[v0.0.5]: https://github.com/typestack/class-sanitizer/compare/v0.0.4...v0.0.5
[v0.0.4]: https://github.com/typestack/class-sanitizer/compare/v0.0.3...v0.0.4
[v0.0.3]: https://github.com/typestack/class-sanitizer/compare/v0.0.1...v0.0.3
[keep-a-changelog]: https://keepachangelog.com/en/1.0.0/
[ct]: https://github.com/typestack/class-transformer
[cv]: https://github.com/typestack/class-validator


src/Sanitizer.ts
import { SanitizeTypes } from './enums';
import { defaultMetadataStorage } from './default-storage.const';
import { CustomSanitizer, SanitationMetadata } from './interfaces';
import validator from 'validator';

/**
 * Sanitizer performs sanitation of the given object based on its metadata.
 */
export class Sanitizer {
  // -------------------------------------------------------------------------
  // Properties
  // -------------------------------------------------------------------------

  private _container: { get(type: Function): any };
  private metadataStorage = defaultMetadataStorage;

  // -------------------------------------------------------------------------
  // Accessors
  // -------------------------------------------------------------------------

  set container(container: { get(type: Function): any }) {
    this._container = container;
  }

  // -------------------------------------------------------------------------
  // Annotation-based Validation Methods
  // -------------------------------------------------------------------------

  /**
   * Performs sanitation of the given object based on annotations used in given object class.
   */
  sanitize(object: any): void {
    this.metadataStorage
      .getSanitizeMetadatasForObject(object.constructor)
      .filter(metadata => !!object[metadata.propertyName])
      .forEach(
        metadata => (object[metadata.propertyName] = this.sanitizeValue(object[metadata.propertyName], metadata))
      );

    // todo: implemented nested sanitation
  }

  /**
   * Performs sanitation of the given object based on annotations used in given object class.
   * Performs in async-style, useful to use it in chained promises.
   */
  sanitizeAsync<T>(object: T): Promise<T> {
    return new Promise<T>(ok => {
      this.sanitize(object);
      ok(object);
    });
  }

  // -------------------------------------------------------------------------
  // Sanitation Methods
  // -------------------------------------------------------------------------

  /**
   * Remove characters that appear in the blacklist. The characters are used in a RegExp and so you will need to
   * escape some chars, e.g @Blacklist('\\[\\]')
   */
  blacklist(str: string, chars: string): string;
  blacklist(str: string, chars: RegExp): string;
  blacklist(str: string, chars: RegExp | string): string {
    return validator.blacklist(str, <string>chars);
  }

  /**
   * Replace <, >, &, ', " and / with HTML entities.
   */
  escape(str: string): string {
    return validator.escape(str);
  }

  /**
   * Trim characters from the left-side of the input.
   */
  ltrim(str: string, chars?: string[]): string {
    return validator.ltrim(str, chars ? chars.join() : undefined);
  }

  /**
   * Canonicalize an email address.
   */
  normalizeEmail(str: string, lowercase?: boolean): string | false {
    return validator.normalizeEmail(str, { all_lowercase: lowercase });
  }

  /**
   * Trim characters from the right-side of the input.
   */
  rtrim(str: string, chars?: string[]): string {
    return validator.rtrim(str, chars ? chars.join() : undefined);
  }

  /**
   * Remove characters with a numerical value < 32 and 127, mostly control characters.
   * If keepNewLines is true, newline characters are preserved (\n and \r, hex 0xA and 0xD).
   * Unicode-safe in JavaScript.
   */
  stripLow(str: string, keepNewLines?: boolean): string {
    return validator.stripLow(str, keepNewLines);
  }

  /**
   * Convert the input to a boolean.
   * Everything except for '0', 'false' and '' returns true. In strict mode only '1' and 'true' return true.
   */
  toBoolean(input: any, isStrict?: boolean): boolean {
    if (typeof input === 'string') {
      return validator.toBoolean(input, isStrict);
    }
    return !!input;
  }

  /**
   * Convert the input to a date, or null if the input is not a date.
   */
  toDate(input: any): Date {
    if (input instanceof Date) {
      return input;
    }
    return validator.toDate(input);
  }

  /**
   * Convert the input to a float.
   */
  toFloat(input: any): number {
    if (typeof input === 'number') {
      return input;
    }
    return validator.toFloat(input);
  }

  /**
   * Convert the input to an integer, or NaN if the input is not an integer.
   */
  toInt(input: any, radix?: number): number {
    if (typeof input === 'number') {
      return input | 0;
    }
    return validator.toInt(input, radix);
  }

  /**
   * Convert the input to a string.
   */
  toString(input: any): string {
    return validator.toString(input);
  }

  /**
   * Trim characters (whitespace by default) from both sides of the input. You can specify chars that should be trimmed.
   */
  trim(str: string, chars?: string[]): string {
    return validator.trim(str, chars ? chars.join() : undefined);
  }

  /**
   * Remove characters that do not appear in the whitelist.
   * The characters are used in a RegExp and so you will need to escape some chars, e.g. whitelist(input, '\\[\\]').
   */
  whitelist(str: string, chars: string): string;
  whitelist(str: string, chars: RegExp): string;
  whitelist(str: string, chars: RegExp | string): string {
    return validator.whitelist(str, <string>chars);
  }

  // -------------------------------------------------------------------------
  // Private Methods
  // -------------------------------------------------------------------------

  private sanitizeValue(value: any, metadata: SanitationMetadata): any {
    switch (metadata.type) {
      case SanitizeTypes.BLACKLIST:
        return this.blacklist(value, metadata.value1);
      case SanitizeTypes.ESCAPE:
        return this.escape(value);
      case SanitizeTypes.LTRIM:
        return this.ltrim(value, metadata.value1);
      case SanitizeTypes.NORMALIZE_EMAIL:
        return this.normalizeEmail(value, metadata.value1);
      case SanitizeTypes.RTRIM:
        return this.rtrim(value, metadata.value1);
      case SanitizeTypes.STRIP_LOW:
        return this.stripLow(value, metadata.value1);
      case SanitizeTypes.TO_BOOLEAN:
        return this.toBoolean(value, metadata.value1);
      case SanitizeTypes.TO_DATE:
        return this.toDate(value);
      case SanitizeTypes.TO_FLOAT:
        return this.toFloat(value);
      case SanitizeTypes.TO_INT:
        return this.toInt(value, metadata.value1);
      case SanitizeTypes.TO_STRING:
        return this.toString(value);
      case SanitizeTypes.TRIM:
        return this.trim(value, metadata.value1);
      case SanitizeTypes.WHITELIST:
        return this.whitelist(value, metadata.value1);
      case SanitizeTypes.CUSTOM_SANITIZATION:
        return this.metadataStorage
          .getSanitizeConstraintsForObject(metadata.value1)
          .map(sanitizerMetadata => {
            if (!sanitizerMetadata.instance) sanitizerMetadata.instance = this.createInstance(sanitizerMetadata.object);

            return sanitizerMetadata.instance;
          })
          .reduce((result, sanitizer) => sanitizer.sanitize(result), value);

      default:
        throw Error(`Wrong sanitation type is supplied ${metadata.type} for value ${value}`);
    }
  }

  private createInstance(object: Function): CustomSanitizer {
    return this._container ? this._container.get(object) : new (<any>object)();
  }
}
